File: Java/src/main/java/at/tugraz/experiments/DataReadNetwork.java
package at.tugraz.experiments;

import at.tugraz.util.MachineInfo;
import at.tugraz.util.Network;
import at.tugraz.runtime.ObjectReader;
import at.tugraz.runtime.ObjectWriter;
import at.tugraz.tweet.TweetStatus;
import at.tugraz.util.CommonThreadPool;
import at.tugraz.util.Const;
import at.tugraz.util.NodeType;
import at.tugraz.util.RootData;
import org.apache.log4j.Logger;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

public class DataReadNetwork {

    private static Logger logger = Logger.getLogger(DataReadNetwork.class);

    public static void main(String[] args) throws IOException, InterruptedException, ExecutionException {

        // disk-to-disk (d2d)
        // disk-to-memory (d2m)
        // memory-to-memory (m2m)
        // memory-to-disk (m2d)

        String inDataPath = System.getProperty("inDataPath");
        String method = System.getProperty("method");
        String config = System.getProperty("config");
        String plan = System.getProperty("plan");

        String localMethod = "Kryo";

        if (plan.equalsIgnoreCase("d2d") || plan.equalsIgnoreCase("d2m"))
            localMethod = method;

        Network network = new Network(config);
        MachineInfo machineInfo = network.getCurrentMachine();
        ObjectReader reader = new ObjectReader(inDataPath, localMethod);

        if (machineInfo.getNodeType() == NodeType.LEAF) {
            ArrayList<RootData> list = new ArrayList<>();
            reader.readObjects(0, machineInfo.getNrow(), list);
            Collections.sort(list);

            ObjectWriter writer = new ObjectWriter(method, machineInfo.getTotalNRow(), Const.NETWORK_PAGESIZE);
            Client client = initClient(machineInfo.getRoot().getIp(), machineInfo.getPort());
            for (RootData rd : list)
                writer.writeObjectToNetworkPage(rd, client.dos, client.dis);

            writer.flushToNetwork(client.dos, client.dis);
            client.socket.close();

        } else if (machineInfo.getNodeType() == NodeType.MIDDLE) {
            ServerSocket serverSocket = new ServerSocket(machineInfo.getPort());
            serverSocket.setSoTimeout(Const.NETWORK_TIMEOUT);
            Client client = initClient(machineInfo.getRoot().getIp(), machineInfo.getPort());

            ArrayList<Task> tasks = new ArrayList<>();
            ArrayList<Task> dataTasks = new ArrayList<>();
            ExecutorService pool = CommonThreadPool.get(machineInfo.getLeaves().size() + 3);
            Socket socket;

            for (int i = 0; i < machineInfo.getLeaves().size(); i++) {
                socket = serverSocket.accept();
                logger.info("Accepted Client In MIDDLE node:" + machineInfo.getIp() + " Client >> " + socket.getLocalAddress() + " - " + socket.getInetAddress().getHostAddress());

                Client cli = new Client(socket, new DataOutputStream(socket.getOutputStream()), new DataInputStream(socket.getInputStream()));
                ObjectReader clientReader = new ObjectReader(method);
                NetworkReadTask clientTask = new NetworkReadTask(cli, clientReader);
                tasks.add(clientTask);
                dataTasks.add(clientTask);
            }

            // read objects from local
            Task clientTask = new LocalReadTask(reader, machineInfo.getNrow());
            tasks.add(clientTask);
            dataTasks.add(clientTask);

            ObjectWriter writer = new ObjectWriter(method, machineInfo.getTotalNRow(), Const.NETWORK_PAGESIZE);
            // add external sort task
            tasks.add(new ExternalSortTask(dataTasks, writer, false, client.dis, client.dos));

            // invoke all tasks
            List<Future<Boolean>> rt = pool.invokeAll(tasks);
            pool.shutdown();

            //check for exceptions
            for (Future<Boolean> f : rt) {
                f.get();
            }
        } else if (machineInfo.getNodeType() == NodeType.ROOT) {
            String outDataPath = System.getProperty("outDataPath");
            ObjectWriter writer = new ObjectWriter(outDataPath, method, machineInfo.getTotalNRow());
            ServerSocket serverSocket = new ServerSocket(machineInfo.getPort());
            serverSocket.setSoTimeout(Const.NETWORK_TIMEOUT);

            ArrayList<Task> tasks = new ArrayList<>();
            ArrayList<Task> dataTasks = new ArrayList<>();
            ExecutorService pool = CommonThreadPool.get(machineInfo.getLeaves().size() + 3);
            Socket socket;
            for (int i = 0; i < machineInfo.getLeaves().size(); i++) {
                socket = serverSocket.accept();
                logger.info("Accepted Client In ROOT node:" + machineInfo.getIp() + " Client >> " + socket.getInetAddress().getHostAddress());

                Client cli = new Client(socket, new DataOutputStream(socket.getOutputStream()), new DataInputStream(socket.getInputStream()));
                ObjectReader clientReader = new ObjectReader(method);
                NetworkReadTask clientTask = new NetworkReadTask(cli, clientReader);
                tasks.add(clientTask);
                dataTasks.add(clientTask);

            }
            // read objects from local
            Task clientTask = new LocalReadTask(reader, machineInfo.getNrow());
            tasks.add(clientTask);
            dataTasks.add(clientTask);

            // add external sort task
            if (plan.equalsIgnoreCase("d2d") || plan.equalsIgnoreCase("m2d"))
                tasks.add(new ExternalSortTask(dataTasks, writer, true, null, null));
            else
                tasks.add(new ExternalSortTask(dataTasks, null, true, null, null));


            List<Future<Boolean>> rt = pool.invokeAll(tasks);
            pool.shutdown();

            //check for exceptions
            for (Future<Boolean> f : rt) {
                f.get();
            }
        }
    }

    private static Client initClient(String ip, int port) throws InterruptedException {
        boolean flag = false;
        Socket socket;
        Client client = null;
        for (int i = 0; i < 10 && !flag; i++) {
            TimeUnit.SECONDS.sleep(1);
            try {
                socket = new Socket(ip, port);
                socket.setSoTimeout(Const.NETWORK_TIMEOUT);
                client = new Client(socket, new DataOutputStream(socket.getOutputStream()), new DataInputStream(socket.getInputStream()));
                flag = socket.isConnected();
            } catch (Exception ignored) {
            }
        }
        if (!flag) throw new RuntimeException("Client can't start >> " + ip + ":" + port);
        else {
           return client;
        }
    }

    private static abstract class Task implements Callable<Boolean> {
        protected final ObjectReader reader;
        protected final BlockingQueue<ArrayList<RootData>> queue;
        protected boolean status;

        public Task(ObjectReader reader) {
            this.reader = reader;
            this.queue = new ArrayBlockingQueue<>(Const.NETWORK_CLIENT_QUEUE_SIZE);
        }
    }

    private static class NetworkReadTask extends Task {
        protected final Client client;

        public NetworkReadTask(Client client, ObjectReader reader) {
            super(reader);
            this.client = client;
        }

        @Override
        public Boolean call() throws IOException, InterruptedException {
            this.status = true;
            while (true) {
                byte ack = 1;
                client.dos.writeByte(ack);
                int pageSize = client.dis.readInt();
                if (pageSize == -1) {
                    client.dos.writeByte(ack);
                    break;
                }
                byte[] buffer = new byte[pageSize];
                int off = 0;
                do {
                    off += client.dis.read(buffer, off, pageSize - off);
                } while (off < pageSize);

                ByteBuffer bb = ByteBuffer.allocate(pageSize).put(buffer);
                ArrayList<RootData> list = new ArrayList<>();
                int relativePosition = 0;
                do {
                    bb.position(relativePosition);
                    int objectSize = bb.getInt();
                    relativePosition += 4;
                    byte[] objectBuffer = new byte[objectSize];
                    bb.position(relativePosition);
                    bb.get(objectBuffer, 0, objectSize);
                    relativePosition += objectSize;
                    list.add(reader.readObjectWithSerialization(new TweetStatus(), objectBuffer));
                } while (relativePosition < pageSize);
                this.queue.put(list);
            }
            this.status = false;
            client.dos.close();
            client.dis.close();
            client.socket.close();
            return false;
        }
    }

    private static class LocalReadTask extends Task {
        protected final int nrow;

        public LocalReadTask(ObjectReader reader, int nrow) {
            super(reader);
            this.nrow = nrow;
        }

        @Override
        public Boolean call() throws IOException, InterruptedException {
            this.status = true;
            ArrayList<RootData> list = new ArrayList<>();
            this.reader.readObjects(0, this.nrow, list);
            Collections.sort(list);

            int chunks = (int) Math.ceil((double) nrow / Const.NETWORK_LOCAL_READ_LENGTH);

            for (int i = 0; i < chunks & i * Const.NETWORK_LOCAL_READ_LENGTH < nrow; i++) {
                ArrayList<RootData> tmpList = new ArrayList<>();
                for (int j = i * Const.NETWORK_LOCAL_READ_LENGTH; j < Math.min((i + 1) * Const.NETWORK_LOCAL_READ_LENGTH, nrow); j++) {
                    tmpList.add(list.get(j));
                }
                this.queue.put(tmpList);
            }
            this.status = false;
            return false;
        }
    }

    private static class Client {
        private final DataInputStream dis;
        private final DataOutputStream dos;
        private final Socket socket;

        public Client(Socket socket, DataOutputStream dos, DataInputStream dis) {
            this.dis = dis;
            this.dos = dos;
            this.socket = socket;
        }

        public DataInputStream getDis() {
            return dis;
        }

        public DataOutputStream getDos() {
            return dos;
        }

        public Socket getSocket() {
            return socket;
        }
    }

    private static class ObjectNetworkIndex implements Comparable<ObjectNetworkIndex> {
        private RootData myObject;
        private int clientIndex;

        public ObjectNetworkIndex(int clientIndex, RootData myObject) {
            this.myObject = myObject;
            this.clientIndex = clientIndex;
        }

        public RootData getValue() {
            return myObject;
        }

        public void setValue(RootData myObject) {
            this.myObject = myObject;
        }

        public int getClientIndex() {
            return clientIndex;
        }

        public void setClientIndex(int clientIndex) {
            this.clientIndex = clientIndex;
        }

        @Override
        public int compareTo(ObjectNetworkIndex o) {
            return this.myObject.compareTo(o.myObject);
        }
    }

    private static class ExternalSortTask extends Task {
        protected final ArrayList<Task> tasks;
        protected final ObjectWriter writer;
        protected final DataInputStream dis;
        protected final DataOutputStream dos;
        protected final boolean onDisk;
        protected final ArrayList<RootData> dataList;

        public ExternalSortTask(ArrayList<Task> tasks, ObjectWriter writer, boolean onDisk, DataInputStream dis, DataOutputStream dos) {
            super(null);
            this.tasks = tasks;
            this.writer = writer;
            this.onDisk = onDisk;
            this.dis = dis;
            this.dos = dos;
            if (writer == null)
                this.dataList = new ArrayList<>();
            else
                this.dataList = null;
        }

        @Override
        public Boolean call() throws InterruptedException {
            int numberOfClients = tasks.size();
            long[] pageObjectCounter = new long[numberOfClients];
            PriorityQueue<ObjectNetworkIndex> queue = new PriorityQueue<ObjectNetworkIndex>();

            // reading objects from the first pages and adding them to a priority queue
            for (int i = 0; i < numberOfClients; i++) {
                List<RootData> listReadFromFile = tasks.get(i).queue.take();
                pageObjectCounter[i] = listReadFromFile.size();
                for (RootData rd : listReadFromFile) {
                    ObjectNetworkIndex objectNetworkIndex = new ObjectNetworkIndex(i, rd);
                    queue.add(objectNetworkIndex);
                }
            }
            logger.info("Network External Sort: First page reading is done! ");

            while (!queue.isEmpty()) {
                ObjectNetworkIndex tmpObjectNetworkIndex = queue.remove();
                int clientNumber = tmpObjectNetworkIndex.getClientIndex();

                // reduce the number of objects read from that file.
                pageObjectCounter[clientNumber]--;

                // If needed load more objects from files.
                // if zero load the next page from file and add objects.
                if (pageObjectCounter[clientNumber] == 0) {
                    List<RootData> listReadFromFile = null;
                    if (!tasks.get(clientNumber).queue.isEmpty()) {
                        listReadFromFile = tasks.get(clientNumber).queue.take();
                    } else {
                        while (tasks.get(clientNumber).status && tasks.get(clientNumber).queue.isEmpty()) ;
                        if (!tasks.get(clientNumber).queue.isEmpty()) {
                            listReadFromFile = tasks.get(clientNumber).queue.take();
                        }
                    }
                    if (listReadFromFile != null) {
                        pageObjectCounter[clientNumber] = listReadFromFile.size();
                        for (RootData rd : listReadFromFile) {
                            ObjectNetworkIndex objectNetworkIndex = new ObjectNetworkIndex(clientNumber, rd);
                            queue.add(objectNetworkIndex);
                        }
                    }
                }
                if (writer !=null) {
                    if (onDisk) writer.writeObjectToFile(tmpObjectNetworkIndex.myObject);
                    else writer.writeObjectToNetworkPage(tmpObjectNetworkIndex.myObject, dos, dis);
                }
                else
                    this.dataList.add(tmpObjectNetworkIndex.myObject);
            }
            logger.info("Network External Sort: Done!");
            if (writer != null) {
                if (onDisk) writer.flush();
                else writer.flushToNetwork(dos, dis);
            }
            return null;
        }
    }
}

File: src/main/rust/src/DataReadNetwork.rs
#![allow(non_snake_case)]
#![allow(dead_code)]
#![allow(unused_variables)]

use std::{io, env};
use std::borrow::Borrow;
use crate::runtime::ObjectReader::ObjectReader;
use crate::runtime::ObjectWriter::ObjectWriter;
use crate::tweetStructs::TweetStatus::TweetStatus;
use priority_queue::PriorityQueue;
use std::cmp::Reverse;
use crate::Const::{NETWORK_PAGESIZE, NodeType};
use crate::util::Const;
use crate::util::ObjectNetworkIndex::ObjectNetworkIndex;
use crate::util::Network::Network;
use std::{thread, time};
use std::net::{TcpListener, TcpStream, Shutdown};
use std::io::{Cursor, Read, Write};
use std::ops::DerefMut;
use std::str::from_utf8;
use bson::{doc, Document};
use byteorder::ReadBytesExt;
use std::sync::atomic::{AtomicUsize, Ordering};
use crossbeam_queue::{ArrayQueue, SegQueue};
use crossbeam_utils::thread::scope;
use rand::{thread_rng, Rng};
use rand::distributions::uniform::SampleBorrow;
use std::sync::{Arc, Mutex};
use std::thread::Thread;


mod tweetStructs;
mod runtime;
mod util;

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    let inDataPath: &str = args[1].as_str();
    let outDataPath: &str = args[2].as_str();
    let method: &str = args[3].as_str();
    let config: String = args[4].to_string();
    let plan: &str = args[5].as_str();

    let mut network = Network::new(config);
    let mut machineInfo = network.getCurrentMachine();
    let mut reader = ObjectReader::new1(inDataPath, method);
    let mut nodeType = machineInfo.getNodeType();
    static mut queues: Vec<ArrayQueue<Vec<TweetStatus>>> = vec![];
    static mut statuses: Vec<bool> = vec![];

    if nodeType == NodeType::LEAF {
        let stream = initClient(machineInfo.root(), machineInfo.port()).ok().unwrap();
        let mut list: Vec<TweetStatus> = vec![];
        reader.readObjects(0, machineInfo.nrow(), &mut list);
        list.sort_by(|cu, ot| cu.getOrder().cmp(&ot.getOrder()));
        let mut writer = ObjectWriter::new2(method, machineInfo.nrow(), Const::NETWORK_PAGESIZE as usize);
        for rd in list {
            writer.writeObjectToNetworkPage(rd, &mut stream.try_clone().unwrap());
        }
        writer.flushToNetwork(&mut stream.try_clone().unwrap());
    } else if nodeType == NodeType::MIDDLE {
        let serverSocket = TcpListener::bind(format!("{}:{}", machineInfo.ip(), machineInfo.port())).unwrap();
        unsafe {
            let mut streams = vec![];

            for i in 0..machineInfo.leaves().len() + 1 {
                let queue = ArrayQueue::new(Const::NETWORK_CLIENT_QUEUE_SIZE);
                queues.push(queue);
                statuses.push(true);
            }
            for i in 0..machineInfo.leaves().len() {
                let stream = serverSocket.incoming().next().unwrap();
                let stream = stream.unwrap();
                streams.push(stream.try_clone().unwrap());
                let method = reader.method().clone();
                thread::spawn(move || {
                    NetworkReadTask(stream, method, queues.get(i).unwrap());
                    statuses[i] = false;
                });
            }
            thread::spawn(move || {
                let index = queues.len() - 1;
                LocalReadTask(&mut reader, queues.get(index).unwrap());
                statuses[index] = false;
            });

            let mut writer = ObjectWriter::new2(method, machineInfo.getTotalNRow(), Const::NETWORK_PAGESIZE as usize);
            let stream = initClient(machineInfo.root(), machineInfo.port()).ok().unwrap();
            ExternalSortTask(&mut queues, &mut statuses, true, writer, false, Some(&mut stream.try_clone().unwrap()));


            let ack = b"1";
            for stream in streams {
                stream.try_clone().unwrap().write(ack);
            }
        }
    } else if nodeType == NodeType::ROOT {
        let serverSocket = TcpListener::bind(format!("{}:{}", machineInfo.ip(), machineInfo.port())).unwrap();
        unsafe {
            let mut streams = vec![];
            for i in 0..machineInfo.leaves().len() + 1 {
                let queue = ArrayQueue::new(Const::NETWORK_CLIENT_QUEUE_SIZE);
                queues.push(queue);
                statuses.push(true);
            }
            for i in 0..machineInfo.leaves().len() {
                let stream = serverSocket.incoming().next().unwrap();
                let stream = stream.unwrap();
                streams.push(stream.try_clone().unwrap());
                let method = reader.method().clone();
                thread::spawn(move || {
                    NetworkReadTask(stream, method, queues.get(i).unwrap());
                    statuses[i] = false;
                });
            }
            thread::spawn(move || {
                let index = queues.len() - 1;
                LocalReadTask(&mut reader, queues.get(index).unwrap());
                statuses[index] = false;
            });

            let writer = ObjectWriter::new1(outDataPath, method, machineInfo.getTotalNRow());
            if plan.to_lowercase().eq("d2d") || plan.to_lowercase().eq("m2d") {
                ExternalSortTask(&mut queues, &statuses, true, writer, true, None);
            } else {
                ExternalSortTask(&mut queues, &statuses, false, writer, false, None);
            }

            let ack = b"1";
            for stream in streams {
                stream.try_clone().unwrap().write(ack);
            }
        }
    }
    Ok(())
}

fn initClient(ip: &str, port: u16) -> Result<TcpStream, Box<dyn std::error::Error>> {
    let delay = time::Duration::from_secs(1);
    for i in 0..1000 {
        match TcpStream::connect(format!("{}:{}", ip, port)) {
            Ok(mut stream) => {
                return Ok(stream);
            }
            _ => { thread::sleep(delay) }
        }
    }

    Err(Box::from(format!("Client can't start >> {}:{}", ip, port)))
}

fn NetworkReadTask(mut stream: TcpStream, method: u16, queue: &ArrayQueue<Vec<TweetStatus>>) {
    let mut i32_data = [0 as u8; 4];
    let ack = b"1";
    let mut c = 0;

    while true {
        stream.write(&ack.clone());
        stream.read_exact(&mut i32_data).unwrap();
        let pageSize = i32::from_be_bytes(i32_data);
        if pageSize == -1 {
            break;
        }
        let mut relativePosition: usize = 0;
        let mut list: Vec<TweetStatus> = vec![];
        while relativePosition < pageSize as usize {
            stream.read_exact(&mut i32_data).unwrap();
            let objectSize = i32::from_be_bytes(i32_data);
            let mut buffer = vec![0u8; objectSize as usize];
            stream.read_exact(&mut buffer);
            let buff_data = buffer.as_slice();

            match method {
                Const::JSON => {
                    list.push(serde_json::from_slice(buff_data).unwrap());
                }
                Const::BINCODE => {
                    list.push(bincode::deserialize(&buff_data).unwrap())
                }
                Const::MESSAGEPACK => {
                    list.push(rmp_serde::from_slice(&buff_data).unwrap())
                }
                Const::BSON => {
                    let doc = Document::from_reader(&mut Cursor::new(&buff_data[..])).unwrap();
                    let bson_data = bson::bson!(doc);
                    list.push(bson::from_bson(bson_data).unwrap())
                }
                Const::FLEXBUF => {
                    list.push(flexbuffers::from_slice(&buff_data).unwrap())
                }
                _ => {
                    println!("The method is not support!!");
                    return;
                }
            }

            relativePosition = relativePosition + 4 + objectSize as usize;
        }
        while queue.is_full() {}
        c +=list.len();
        queue.push(list);

    }

    println!(" c={}", c);
}

fn LocalReadTask(reader: &mut ObjectReader, queue: &ArrayQueue<Vec<TweetStatus>>) {
    let mut list: Vec<TweetStatus> = vec![];
    let nrow = reader.getRlen();
    reader.readObjects(0, nrow, &mut list);
    list.sort_by(|cu, ot| cu.getOrder().cmp(&ot.getOrder()));
    let mut c = 0;
    for ch in list.chunks_mut(Const::NETWORK_LOCAL_READ_LENGTH as usize) {
        while queue.is_full() {}
        let tmp = ch.to_vec();
        c += tmp.len();
        queue.push(tmp);
    }
    println!("local c={}",c);
}

fn ExternalSortTask(queues: &mut Vec<ArrayQueue<Vec<TweetStatus>>>, statuses: &Vec<bool>, is_write: bool, mut writer: ObjectWriter, onDisk: bool, stream: Option<&TcpStream>) {
    let mut dataList: Vec<TweetStatus> = vec![];
    let numberOfClients = queues.len();
    let mut pageObjectCounter: Vec<u64> = vec![0; numberOfClients];
    let mut queue: PriorityQueue<ObjectNetworkIndex, Reverse<usize>> = PriorityQueue::new();

    // reading objects from the first pages and adding them to a priority queue
    for i in 0..numberOfClients as u32 {
        while queues[i as usize].is_empty() {}
        let mut listReadFromFile = queues[i as usize].pop().unwrap();
        pageObjectCounter[i as usize] = listReadFromFile.len() as u64;
        for rd in listReadFromFile {
            let order = rd.getOrder();
            let objectNetworkIndex = ObjectNetworkIndex::new(rd, i);
            queue.push(objectNetworkIndex, Reverse(order));
        }
    }
    println!("Network External Sort: First page reading is done! ");
    let mut c = 0;
    let mut c1 = 0;
    let mut c2 = 0;
    let mut c3 = 0;
    while !queue.is_empty() {
        let tmpObjectNetworkIndex: ObjectNetworkIndex = queue.pop().unwrap().0;
        let clientNumber = tmpObjectNetworkIndex.getClientIndex() as usize;

        // reduce the number of objects read from that file.
        pageObjectCounter[clientNumber] = pageObjectCounter[clientNumber] - 1;

        // If needed load more objects from files.
        // if zero load the next page from file and add objects.
        if pageObjectCounter[clientNumber] == 0 {
            let mut listReadFromFile: Vec<TweetStatus> = vec![];
            if !queues[clientNumber].is_empty() {
                listReadFromFile = queues[clientNumber].pop().unwrap();
            } else {
                while statuses[clientNumber] && queues[clientNumber].is_empty() {}
                if !queues[clientNumber].is_empty() {
                    listReadFromFile = queues[clientNumber].pop().unwrap();
                }
            }
            if listReadFromFile.len() > 0 {
                pageObjectCounter[clientNumber] = listReadFromFile.len() as u64;
                for rd in listReadFromFile {
                    let order = rd.getOrder();
                    let objectNetworkIndex = ObjectNetworkIndex::new(rd, clientNumber as u32);
                    queue.push(objectNetworkIndex, Reverse(order));

                }
            }
        }
        if is_write {
            if onDisk { writer.writeObjectToFile(&tmpObjectNetworkIndex.getObject()); } else {
                writer.writeObjectToNetworkPage(tmpObjectNetworkIndex.getObject(), &mut Option::from(stream.unwrap().try_clone()).unwrap().unwrap());
            }
        } else {
            dataList.push(tmpObjectNetworkIndex.getObject());
        }
        c +=1;
        if clientNumber == 0 {
            c1 +=1;
        }
        if clientNumber == 1 {
            c2 +=1;
        }
        if clientNumber == 2 {
            c3 +=1;
        }
    }

    println!("Network External Sort: Done! c={}  c1={}  c2={}  c3={}  q1={} q1c={} q2={} q1c={} q3={} q1c={}", c, c1, c2, c3, queues[0].len(), queues[0].capacity(),
             queues[1].len(), queues[1].capacity(),queues[2].len(), queues[2].capacity(),);
    if is_write {
        if onDisk { writer.flush(); } else {
            writer.flushToNetwork(&mut Option::from(stream.unwrap().try_clone()).unwrap().unwrap());
        }
    }
}